pkg/vault/contracts/BatchRouterHooks.sol

```diff
@@ -378,163 +429,237 @@ contract BatchRouterHooks is BatchRouterCommon {
+            // to the Router, which acts as an intermediary. If the sender is the Router, we just skip this
+            // step (useful for queries).
+            _permit2.transferFrom(sender, address(this), stepMaxAmountIn.toUint160(), address(stepTokenIn));
+        }
+
+        (uint256[] memory exactAmountsOut, ) = _getSingleInputArrayAndTokenIndex(pool, tokenOut, stepExactAmountOut);
+
+        // Router is always an intermediary in this case. The Vault will burn tokens from the Router, so
+        // Router is both owner and spender (which doesn't need approval).
+        (amountIn, , ) = _vault.removeLiquidity(
+            RemoveLiquidityParams({
+                pool: pool,
+                from: address(this),
+                maxBptAmountIn: stepMaxAmountIn,
+                minAmountsOut: exactAmountsOut,
+                kind: RemoveLiquidityKind.SINGLE_TOKEN_EXACT_OUT,
+                userData: userData
+            })
+        );
+
+        if (isLastStep) {
+            // Refund unused portion of BPT to the user.
+            if (amountIn < stepMaxAmountIn && sender != address(this)) {
+                stepTokenIn.safeTransfer(address(sender), stepMaxAmountIn - amountIn);
+            }
+
+            _settledTokenAmounts().tAdd(address(stepTokenIn), amountIn);
+        } else {
+            if (amountIn < stepMaxAmountIn) {
+                // Refund unused portion of BPT flashloan to the Vault.
+                uint256 refundAmount = stepMaxAmountIn - amountIn;
+                stepTokenIn.safeTransfer(address(_vault), refundAmount);
+                _vault.settle(stepTokenIn, refundAmount);
+            }
+        }
+    }
+
+    function _addLiquidityExactOut(
+        bytes memory userData,
+        address sender,
+        address pool,
+        IERC20 stepTokenIn,
+        IERC20 tokenOut,
+        uint256 stepExactAmountOut,
+        uint256 stepMaxAmountIn,
+        bool isFirstStep,
+        bool isLastStep
+    ) internal returns (uint256 amountIn) {
+        // Token out is BPT: add liquidity - Single token exact out.
+        (uint256[] memory stepAmountsIn, uint256 tokenIndex) = _getSingleInputArrayAndTokenIndex(
+            pool,
+            stepTokenIn,
+            stepMaxAmountIn
+        );
+
+        // Reusing `amountsIn` as input argument and function output to prevent stack too deep error.
+        (stepAmountsIn, , ) = _vault.addLiquidity(
+            AddLiquidityParams({
+                pool: pool,
+                to: isFirstStep ? sender : address(_vault),
+                maxAmountsIn: stepAmountsIn,
+                minBptAmountOut: stepExactAmountOut,
+                kind: AddLiquidityKind.SINGLE_TOKEN_EXACT_OUT,
+                userData: userData
+            })
+        );
+
+        amountIn = stepAmountsIn[tokenIndex];
+
+        if (isLastStep) {
+            _currentSwapTokenInAmounts().tAdd(address(stepTokenIn), amountIn);
+        } else {
+            stepExactAmountOut = amountIn;
```

Comment:
```suggestion
             // Note that this assignment does not affect the caller's `stepExactAmountOut`.
            stepExactAmountOut = amountIn;
```
I think this is ok; this local assignment affects the next statement, but there's also an "outer" stepExactAmountOut which is shadowed here, and unaffected. Double check that it isn't used later (doesn't seem to be; the amountIn assignment seems to cover it), and maybe consider renaming it in this function to localStepExactAmountOut or stepExactAmountOutParam to be completely clear.